apis/python/src/tiledbsoma/io.py

```diff
@@ -477,20 +477,43 @@ def add_X_layer(
     exp: Experiment,
     measurement_name: str,
     X_layer_name: str,
-    X_layer_data: Union[
-        Matrix, h5py.Dataset
-    ],  # e.g. a scipy.csr_matrix from scanpy analysis
+    # E.g. a scipy.csr_matrix from scanpy analysis:
+    X_layer_data: Union[Matrix, h5py.Dataset],
     ingest_mode: IngestMode = "write",
 ) -> None:
     """
     This is useful for adding X data, for example from scanpy.pp.normalize_total, scanpy.pp.log1p, etc.
 
     Use `ingest_mode="resume"` to not error out if the schema already exists.
     """
-    uri = f"{exp.ms[measurement_name].X.uri}/{X_layer_name}"
-    sparse_nd_array = SparseNDArray(uri)
-    create_from_matrix(sparse_nd_array, X_layer_data, ingest_mode=ingest_mode)
-    exp.ms[measurement_name].X.set(X_layer_name, sparse_nd_array)
+    add_matrix_to_collection(exp, measurement_name, "X", X_layer_name, X_layer_data)
+
+
+def add_matrix_to_collection(
+    exp: Experiment,
+    measurement_name: str,
+    collection_name: str,
+    matrix_name: str,
+    # E.g. a scipy.csr_matrix from scanpy analysis:
+    matrix_data: Union[Matrix, h5py.Dataset],
+    ingest_mode: IngestMode = "write",
+) -> None:
+    """
+    This is useful for adding X/obsp/varm/etc data, for example from scanpy.pp.normalize_total,
+    scanpy.pp.log1p, etc.
+
+    Use `ingest_mode="resume"` to not error out if the schema already exists.
+    """
+    cls = (
+        DenseNDArray
+        if isinstance(matrix_data, (np.ndarray, h5py.Dataset))
+        else SparseNDArray
+    )
+    collection = cast(Collection, exp.ms[measurement_name][collection_name])
```

Comment:
needs some handling for the case where this is the _first_ `obsm` etc -- this code (as currently written) only handles adding _more_ when the collection already exists